### ISSUE TITLE
proposal: no constructor should give const suggestion?

### ISSUE THREAD

**Issue by stephane-archer** [dart-lang/language#3965]
moving from:
```
class _FalseColorSwitch extends StatelessWidget {

  @override
  Widget build(BuildContext context) {
``` 
to 
```
class _FalseColorSwitch extends StatelessWidget {
  const _FalseColorSwitch(); // adding a const constructor that does nothing

  @override
  Widget build(BuildContext context) {
``` 
give me so many linter optimization suggestions:
```
Use 'const' with the constructor to improve performance.
Try adding the 'const' keyword to the constructor invocation.
``` 

I think no constructor should give me the same recommendation as a const constructor that does nothing. is there an actual difference?

**Comment by srawlins** [comment]
Moving over to the language repo for the language request. I think this has been asked and answered before ("const by default"), but someone here can point to the past response.

**Comment by mateusfccp** [comment]
Refer to #1410. I don't know if it's the exact same case, but it has a related discussion explaining how some things work.

**Comment by lrhn** [comment]
The reason you don't get `const` without asking for it, is that you might get `const` without wanting it.

A constructor being `const` is a *promise* to always allow `const` construction of the class. Changing that is a breaking change.
That kind of promise is not something the system should make for you, it's something you have to explicitly state.

If not, you could accidentally have a class that supports `const` construction in one version of a package, and not in the next one, because you added a mutable cache for some variable, and never even considered that the class could be `const`, because it was never intended as such. (So it's not tested.)
And that could break someone else, who relied on the implicit `const`-ness of the former version.

(I have no opinion on what lints should suggest here, but the default constructor you get if you write nothing is not a `const` constructor, whether it could be or not.)

**Comment by stephane-archer** [comment]
@lrhn I understand now why the constructor is not `const` by default if it can.
I think a linter suggestion to add a const constructor would be nice.
That way people don't miss the opportunity for extra optimization and it's explicit that they have a `const` constructor, so the breaking change will be clear to anyone (especially removing it).

**Comment by lrhn** [comment]
Ack. That's why I consider this an analyzer or linter issue, not a language issue.

**Comment by stephane-archer** [comment]
@lrhn I agree, that is why it was originally in `dart-lang/linter` but @srawlins transferred it here.
Could you please transfer it back?
